src/model/guild/automod.rs

```diff
@@ -0,0 +1,859 @@
+#[derive(Clone, Debug, PartialEq)]
+pub enum Action {
+    /// Blocks the content of a message according to the rule.
+    BlockMessage,
+    /// Logs user content to a specified channel.
+    Alert(ChannelId),
+    /// Timeout user for a specified duration.
+    ///
+    /// Maximum of 2419200 seconds (4 weeks).
+    ///
+    /// A `Timeout` action can only be setup for [`Keyword`] rules.
+    /// [`Permissions::MODERATE_MEMBERS`] permission is required to use the `Timeout` action type.
+    ///
+    /// [`Keyword`]: TriggerType::Keyword
+    /// [`Permissions::MODERATE_MEMBERS`]: crate::model::Permissions::MODERATE_MEMBERS
+    Timeout(u64),
+    Unknown(u8),
+}
+
+/// Gateway event payload sent when a rule is triggered and an action is executed (e.g. message is
+/// blocked).
+///
+/// [Discord docs](https://discord.com/developers/docs/topics/gateway#auto-moderation-action-execution)
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct ActionExecution {
+    /// ID of the guild in which the action was executed.
+    pub guild_id: GuildId,
+    /// Action which was executed.
+    pub action: Action,
+    /// ID of the rule which action belongs to.
+    pub rule_id: RuleId,
+    /// Trigger type of rule which was triggered.
+    #[serde(rename = "rule_trigger_type")]
+    pub trigger_type: TriggerType,
+    /// ID of the user which generated the content which triggered the rule.
+    pub user_id: UserId,
+    /// ID of the channel in which user content was posted.
+    pub channel_id: Option<ChannelId>,
+    /// ID of any user message which content belongs to.
+    ///
+    /// Will be `None` if message was blocked by automod or content was not part of any message.
+    pub message_id: Option<MessageId>,
+    /// ID of any system auto moderation messages posted as a result of this action.
+    ///
+    /// Will be `None` if this event does not correspond to an action with type [`Action::Alert`].
+    pub alert_system_message_id: Option<MessageId>,
+    /// User generated text content.
+    ///
+    /// Requires [`GatewayIntents::MESSAGE_CONTENT`] to receive non-empty values.
+    ///
+    /// [`GatewayIntents::MESSAGE_CONTENT`]: crate::model::gateway::GatewayIntents::MESSAGE_CONTENT
+    pub content: String,
+    /// Word or phrase configured in the rule that triggered the rule.
+    pub matched_keyword: Option<String>,
+    /// Substring in content that triggered the rule.
+    ///
+    /// Requires [`GatewayIntents::MESSAGE_CONTENT`] to receive non-empty values.
+    ///
+    /// [`GatewayIntents::MESSAGE_CONTENT`]: crate::model::gateway::GatewayIntents::MESSAGE_CONTENT
+    pub matched_content: Option<String>,
+}
+
+#[derive(Deserialize, Serialize)]
+#[serde(rename = "ActionMetadata")]
+struct Alert {
+    channel_id: ChannelId,
+}
+
+#[derive(Deserialize, Serialize)]
+#[serde(rename = "ActionMetadata")]
+struct Timeout {
+    #[serde(rename = "duration_seconds")]
+    duration: u64,
+}
+
+impl<'de> Deserialize<'de> for Action {
```

Comment:
Why manual Deserialize :upside_down_face: ?

Reply:
serde supports only strings for [internally tagged] enums then it could be replace the `#[serde(tag = "type")]`.

[internally tagged]: https://serde.rs/enum-representations.html#internally-tagged

Reply:
Go ahead